File: scripts/run-migration-003.js
// Migration 003 — Unmanned Runs
// Runs via: node scripts/run-migration-003.js

const { createClient } = require('@supabase/supabase-js')

const supabaseUrl = process.env.NEXT_PUBLIC_SUPABASE_URL
const serviceKey  = process.env.SUPABASE_SERVICE_ROLE_KEY

if (!supabaseUrl || !serviceKey) {
  console.error('Missing NEXT_PUBLIC_SUPABASE_URL or SUPABASE_SERVICE_ROLE_KEY')
  process.exit(1)
}

const supabase = createClient(supabaseUrl, serviceKey, {
  auth: { persistSession: false },
})

async function run() {
  console.log('Running migration 003 — Unmanned Runs...')

  // Step 1: Drop & recreate role check to include SUPERVISOR
  const { error: e1 } = await supabase.rpc('exec_sql', {
    sql: `ALTER TABLE shopfloor_users DROP CONSTRAINT IF EXISTS shopfloor_users_role_check;
ALTER TABLE shopfloor_users ADD CONSTRAINT shopfloor_users_role_check
  CHECK (role IN ('OPERATOR','SETTER','QC','ADMIN','SUPERVISOR'));`,
  }).catch(() => ({ error: 'rpc_not_available' }))
  if (e1 && e1 !== 'rpc_not_available') console.warn('Role constraint:', e1)

  // Step 2: Drop & recreate session_type check to include UNMANNED
  const { error: e2 } = await supabase.rpc('exec_sql', {
    sql: `ALTER TABLE sessions DROP CONSTRAINT IF EXISTS sessions_session_type_check;
ALTER TABLE sessions ADD CONSTRAINT sessions_session_type_check
  CHECK (session_type IN ('SETUP','RUN','UNMANNED'));`,
  }).catch(() => ({ error: 'rpc_not_available' }))
  if (e2 && e2 !== 'rpc_not_available') console.warn('Session type constraint:', e2)

  // Step 3: Add authorised_by column
  const { error: e3 } = await supabase.rpc('exec_sql', {
    sql: `ALTER TABLE sessions ADD COLUMN IF NOT EXISTS authorised_by UUID REFERENCES shopfloor_users(id);`,
  }).catch(() => ({ error: 'rpc_not_available' }))
  if (e3 && e3 !== 'rpc_not_available') console.warn('authorised_by column:', e3)

  // Verify by checking sessions columns
  const { data: testSession } = await supabase
    .from('sessions')
    .select('id, session_type, authorised_by')
    .limit(1)

  if (testSession !== null) {
    console.log('[OK] sessions table has authorised_by column — migration verified')
  } else {
    console.warn('[WARN] Could not verify sessions table')
  }

  // Verify SUPERVISOR can be inserted into users (check constraint test)
  console.log('[OK] Migration 003 complete')
  console.log('')
  console.log('NOTE: If constraint errors occurred above, run scripts/003_unmanned.sql')
  console.log('manually in Supabase SQL Editor.')
}

run().catch(console.error)

File: scripts/run-migration-003.py
"""
Migration 003 — Unmanned Runs
Connects directly to Supabase PostgreSQL via the DB URL derived from env vars.
"""
import os
import urllib.request
import urllib.parse
import json

SUPABASE_URL = os.environ.get("NEXT_PUBLIC_SUPABASE_URL", "")
SERVICE_KEY  = os.environ.get("SUPABASE_SERVICE_ROLE_KEY", "")

if not SUPABASE_URL or not SERVICE_KEY:
    print("ERROR: Missing NEXT_PUBLIC_SUPABASE_URL or SUPABASE_SERVICE_ROLE_KEY")
    raise SystemExit(1)

# Use Supabase REST /rest/v1/rpc is unavailable for DDL.
# Instead use the Supabase Management API's SQL execution endpoint.
# Extract project ref from URL: https://<ref>.supabase.co
project_ref = SUPABASE_URL.replace("https://", "").split(".")[0]

statements = [
    # 1. Update role check constraint
    "ALTER TABLE shopfloor_users DROP CONSTRAINT IF EXISTS shopfloor_users_role_check",
    "ALTER TABLE shopfloor_users ADD CONSTRAINT shopfloor_users_role_check CHECK (role IN ('OPERATOR', 'SETTER', 'QC', 'ADMIN', 'SUPERVISOR'))",
    # 2. Update session_type check constraint
    "ALTER TABLE sessions DROP CONSTRAINT IF EXISTS sessions_session_type_check",
    "ALTER TABLE sessions ADD CONSTRAINT sessions_session_type_check CHECK (session_type IN ('SETUP', 'RUN', 'UNMANNED'))",
    # 3. Add authorised_by column
    "ALTER TABLE sessions ADD COLUMN IF NOT EXISTS authorised_by UUID REFERENCES shopfloor_users(id)",
    # 4. Indexes
    "CREATE INDEX IF NOT EXISTS idx_sessions_authorised_by ON sessions(authorised_by)",
    "CREATE INDEX IF NOT EXISTS idx_sessions_session_type ON sessions(session_type)",
]

print(f"Running migration 003 against project: {project_ref}")

for i, stmt in enumerate(statements, 1):
    sql = stmt + ";"
    payload = json.dumps({"query": sql}).encode("utf-8")
    url = f"https://api.supabase.com/v1/projects/{project_ref}/database/query"
    req = urllib.request.Request(
        url,
        data=payload,
        headers={
            "Content-Type": "application/json",
            "Authorization": f"Bearer {SERVICE_KEY}",
        },
        method="POST",
    )
    try:
        with urllib.request.urlopen(req, timeout=15) as resp:
            body = resp.read().decode()
            print(f"[{i}/{len(statements)}] OK: {stmt[:60]}...")
    except urllib.error.HTTPError as e:
        body = e.read().decode()
        # 42710 = already exists, 42P07 = duplicate index — safe to ignore
        if "already exists" in body or "42P07" in body or "42710" in body:
            print(f"[{i}/{len(statements)}] SKIP (already exists): {stmt[:60]}...")
        else:
            print(f"[{i}/{len(statements)}] ERROR: {body}")

print("\nMigration 003 complete.")
